keps/sig-storage/20190129-csi-volume-resizing.md

```diff
@@ -0,0 +1,170 @@
+title: Support for CSI volume resizing
+authors:
+  - "@gnufied
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+reviewers:
+  - @saad-ali
+  - @jsafrane
+approvers:
+  - @saad-ali
+  - @childsb
+creation-date: 2019-01-29
+last-updated: 2019-01-29
+status: implementable
+see-also:
+    - [Kubernetes Volume expansion](https://github.com/kubernetes/enhancements/issues/284)
+    - [Online resizing design](https://github.com/kubernetes/enhancements/pull/737)
+replaces:
+superseded-by:
+---
+
+# Support for CSI volume resizing
+
+## Table of Contents
+
+Table of Contents
+=================
+
+   * [Support for CSI volume resizing](#support-for-csi-volume-resizing)
+      * [Table of Contents](#table-of-contents)
+   * [Table of Contents](#table-of-contents-1)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [External resize controller](#external-resize-controller)
+         * [Expansion on Kubelet](#expansion-on-kubelet)
+             * [Offline volume resizing on kubelet:](#offline-volume-resizing-on-kubelet)
+             * [Online volume resizing on kubelet:](#online-volume-resizing-on-kubelet)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Test Plan](#test-plan)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+
+## Summary
+
+To bring CSI volumes in feature parity with in-tree volumes we need to implement support for resizing of CSI volumes.
+
+## Motivation
+
+We recently implemented volume resizing support in CSI specs. This proposal implements this feature for Kubernetes.
+Any CSI volume plugin that implements necessary part of CSI specs will become resizable.
+
+### Goals
+
+To enable expansion of CSI volumes used by `PersistentVolumeClaim`s that support volume expansion as a plugin capability.
+
+### Non-Goals
+
+The expansion capability of a CSI plugin will not be validated by using CSI RPC call when user edits the PVC(i.e existing resize admission controller will not make CSI RPC call).
+The responsibility of
+actually enabling expansion for certains storageclasses still falls on Kubernetes admin.
+
+## Proposal
+
+The design of CSI volume resizing is made of two parts.
+
+
+### External resize controller
+
+To support resizing of CSI volumes an external resize controller will monitor all PVCs.  If a PVC meets following criteria for resizing, it will be added to
+controller's workqueue:
+
+- The driver name disovered from PVC should match name of driver currently known(by querying driver info via CSI RPC call) to external resize controller.
+- Once it notices a PVC has been updated and by comparing old and new PVC object, it determines more space has been requested by the user.
+
+Once PVC gets picked from workqueue, the controller will also compare requested PVC size with actual size of volume in `PersistentVolume`
```

Comment:
We also need to check PV size before actually doing controller resize because it is possible that, volume already has been resized and controller is just resyncing PVCs.